### PR TITLE
McaFee ATD - fixed add malicious, long TaskIds, 401 error on multiple api calls in 1 command, test playbook

### PR DESCRIPTION
**Pull request by bakatzir**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/13926
fixes: https://github.com/demisto/etc/issues/13896
fixes: https://github.com/demisto/etc/issues/13942
## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
screen shot 2018-12-03 at 10 25 43

screen shot 2018-12-03 at 11 27 34

## Does it break backward compatibility?
   - No

## Must have
- [x] Tests
- [x] Documentation (with link to it)
- [ ] Code Review

## Additional changes
Describe additional changes done, for example adding a function to common server.

### CODE REVIEW

File: Integrations/integration-ATD.yml [hunk 1]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            'ContentsFormat': formats['json'],
+            'Contents': res['results'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': {
+                  "ATD.Session(val.SessionId==obj.SessionId)": prettify_current_user_res(res['results'])
             }
+        })
+
+    def prettify_list_users_res(users):
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Shouldn't be in helper functions section?

File: Integrations/integration-ATD.yml [hunk 2]
```diff
@@ -37,672 +38,670 @@ configuration:
...
-
-    function sendRequest(uri, method, headers, body){
-        body = sendRequestRaw(uri,method, headers, body).Body;
-        var res = JSON.parse(body);
-        if (res.success) {
-            if (res.success === 'false') {
-                throw 'ATD Api call to '+uri+' failed with request body: '+body;
+        return prettyCurrentUser
+
+    def get_session():
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Shouldn't be in helper functions section?

File: Integrations/integration-ATD.yml [hunk 3]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+
+
+
+    ''' FUNCTIONS '''
+    def heart_beat():
+        return http_request('php/heartbeat.php','get', API_HEADERS, HEARTBEATHEADERS)
+
+    def test_get_session():
+        res = http_request('php/session.php','get', LOGINHEADERS)
+
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Shouldn't be in helper functions section?

File: Integrations/integration-ATD.yml [hunk 4]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            result = json.loads(result)
+            if 'success' in result:
+                if (result['success'] == 'false'):
+                    return_error('ATD Api call to ' + uri + ' failed with request body: ' + body)
+        return result
+
+
+
+    ''' FUNCTIONS '''
+    def heart_beat():
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Shouldn't be in helper functions section?

File: Integrations/integration-ATD.yml [hunk 5]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+    def list_users(userType):
+        userType = userType if userType else 'STAND_ALONE'
+        result = http_request('php/briefUserList.php?userType='+userType,'get', API_HEADERS)
+        users = result['results']
+        return users
+
+    def list_users_command():
+        users = list_users(demisto.args()['userType'])
+
+        prettyUsers = prettify_list_users_res(users)
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Please handle the case no results found, by returning a string and not an empty table.

File: Integrations/integration-ATD.yml [hunk 6]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            'ContentsFormat': formats['json'],
+            'Contents': users,
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': {
+                  "ATD.Users(val.UserId==obj.UserId)": prettyUsers,
+            }
+        })
+
+    def prettify_list_profiles_res(profiles):
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Shouldn't be in helper functions section?

File: Integrations/integration-ATD.yml [hunk 7]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+                'Name': profiles[i]['name'],
+                'AnalyzerProfileId': profiles[i]['vmProfileid'],
+                'Description': profiles[i]['vmDesc'],
+                'Sandbox': 'True' if profiles[i]['sandbox'] == 1 else 'False',
+                'Internet': 'True' if profiles[i]['internet'] == 1 else 'False',
+                'LocalBlackList': 'True' if profiles[i]['locBlackList'] == 1 else 'False'
+            })
+        return prettyProfiles
+
+    def list_profiles():
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Please split to command and requests functions.

File: Integrations/integration-ATD.yml [hunk 8]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+
+
+        demisto.results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': res['results'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': {
+                  "ATD.ListAnalyzerProfiles(val.AnalyzerProfileId==obj.AnalyzerProfileId)": prettify_list_profiles_res(res['results'])
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: `==` should be `===`

File: Integrations/integration-ATD.yml [hunk 9]
```diff
@@ -37,672 +38,670 @@ configuration:
...
-            Type: entryTypes.note,
-            Contents: users,
-            ContentsFormat: formats.json,
-            HumanReadable: md,
-            EntryContext: {
-                "ATD.Users(val.UserId==obj.UserId)": prettyUsers,
+    def check_task_status_by_jobId(jobIds):
+        taskIds = []
+        for i in range(len(jobIds)):
+            taskIds.append((http_request('php/getTaskIdList.php?jobId='+jobIds[i],'get', API_HEADERS))['result']['taskIdList'])
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: split this line into a few to be easier to read

File: Integrations/integration-ATD.yml [hunk 10]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+
+    def check_task_status_command():
+        ids = []
+        result = {}
+        args = demisto.args()
+
+        if ( ('jobId' not in args and 'taskId' not in args) or ('jobId' in args and 'taskId' in args) ):
+            return_error('You must specify one (and only one) of the following: jobId, taskId.')
+
+        if ('jobId' in args):
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: no need in brackets

File: Integrations/integration-ATD.yml [hunk 11]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+        )
+
+        demisto.results({
+            'Type': entryTypes['note'],
+            'ContentsFormat': formats['json'],
+            'Contents': result['multipleResults'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': {
+                  "ATD.status": result['status'], #backward compatability
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: `"` -> `'`

File: Integrations/integration-ATD.yml [hunk 12]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            multipleMd.append({
+                'taskId': results[i]['result']['taskIdList'],
+                'jobId': jobIds[i]
+            })
+            ec.append({
+                'taskId': results[i]['result']['taskIdList'],
+                'jobId': jobIds[i]
+            })
+
+        md = tblToMd('ATD TaskIds and JobIds List', multipleMd, ['taskId', 'jobId'])
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Please use `tableToMarkdown` (easier for people who are not familiar with our code to understand)

File: Integrations/integration-ATD.yml [hunk 13]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            file_up = {'amas_filename': f}
+            res = http_request(
+                uri,
+                'post',
+                API_HEADERS,
+                body,
+                '',
+                files=file_up,
+                )
+
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: lynt properly

File: Integrations/integration-ATD.yml [hunk 14]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+        if (res['success'] == False):
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Best practise is `if not res['success']:`

File: Integrations/integration-ATD.yml [hunk 15]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            "ATD.taskId": result['taskId'] # backward compatabillity
         }
-        return checkTaskStatusByTaskId(taskIds);
-    }
-
-    function checkTaskStatusCommand(){
-        var ids = [];
-        var result;
+        if 'url' in args:
+            upload_file_output["URL(val.Data == obj.Data).Data"] = sample
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Use our constant paths from common server

File: Integrations/integration-ATD.yml [hunk 16]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            'ContentsFormat': formats['json'],
+            'Contents': result['resultObj'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': md,
+            'EntryContext': upload_file_output
+        })
+
+    def build_report_context(reportSummary, uploadData, status, threshold):
+        context = {}
+        if ((reportSummary) and (reportSummary['Subject'])):
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: please remove unnecessary brackets

File: Integrations/integration-ATD.yml [hunk 17]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+            subject = reportSummary['Subject']
+            context = { 'DBotScore': {'Vendor': 'McAfee-Advanced-Threat-Defense', 'Score': 0 }}
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: please lynt properly

File: Integrations/integration-ATD.yml [hunk 18]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+                    if subject['Type'] == 'application/url':
+                        context['URL(val.Name == obj.Data)'] = {
+                            'Type' : subject['Type'],
+                            'MD5' : subject['md5'],
+                            'SHA1' : subject['sha-1'],
+                            'SHA256' : subject['sha-256'],
+                            'Size' : subject['size'],
+                            'Name' : subject['Name'],
+                            'Malicious': {
+                                'Vendor': 'McAfee-Advanced-Threat-Defense',
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: What are the dashes for?

File: Integrations/integration-ATD.yml [hunk 19]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+    def get_report(uriSuffix, filename, report_type, uploadData, status, threshold):
+        jres = json.loads(http_request('php/showreport.php?' + uriSuffix + '&iType=json', 'get', API_HEADERS))
+        summary = jres['Summary']
+        summary['VerdictDescription'] = summary['Verdict']['Description']
+        summary['VerdictSeverity'] = summary['Verdict']['Severity']
+        ec = build_report_context(summary, uploadData, status, threshold)
+        jresString = json.dumps(jres)
+        if (report_type == 'json'):
+            md = tableToMarkdown('McAfee ATD Sandbox Report', summary, summary.keys(), None, True)
+            return{
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: add space after `return`

File: Integrations/integration-ATD.yml [hunk 20]
```diff
@@ -37,672 +38,670 @@ configuration:
...
-            'srcIp': fileUploadRes.results[0].srcIp,
-            'destIp': fileUploadRes.results[0].destIp,
-            'MD5': fileUploadRes.results[0].md5,
-            'SHA1': fileUploadRes.results[0].sha1,
-            'SHA256': fileUploadRes.results[0].sha256,
-        }
-        return prettyFileUpload;
-    }
+    def detonate(submitType ,sample, timeout, report_type, threshold, fileName):
+        sys.exit(0)
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: This doesn't make much sense..

File: Integrations/integration-ATD.yml [hunk 21]
```diff
@@ -37,672 +38,670 @@ configuration:
...
-        case 'detonate-url':
-            return detonate(1, args.url, args.timeout, args.format, args.threshold, args.fileName);
-            //submit type for url submission is 1
-        //case 'detonate-file-remote':
-            //return detonate(3, args.url, args.timeout, args.format, args.threshold);
-            //submit type for url-download is 3
-        default:
-    }
-  type: javascript
+    logout()
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: should be inside `finally:`

File: Integrations/integration-ATD.yml [hunk 22]
```diff
@@ -998,28 +1024,28 @@ script:
     - name: fileName
       description: The name of the file
     outputs:
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Make sure CS is aware of this change as it breaks BC

File: Integrations/integration-ATD.yml [hunk 23]
```diff
@@ -1069,10 +1095,7 @@ script:
     - contextPath: IP.Address
       description: IP's relevant to the sample
       type: string
-    - contextPath: ATD.Task.IsComplete
-      description: Stating if the detonation was complete or not
-      type: string
-    description: Detonate file through McAfee ATD
+    description: Deprecated, please use detonate playbook instead.
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Remove `please` please :)
and if it's deprecated, then please add `deprecate: true`

File: Integrations/integration-ATD.yml [hunk 24]
```diff
@@ -1107,28 +1130,28 @@ script:
...
-    - contextPath: File.SHA256
+    - contextPath: InfoFile.SHA256
       description: SHA256 hash of the file (only in case of report type=json)
       type: string
-    - contextPath: File.Malicious.Vendor
+    - contextPath: InfoFile.Malicious.Vendor
       description: For malicious files, the vendor that made the decision
       type: string
-    - contextPath: File.Malicious.Description
+    - contextPath: InfoFile.Malicious.Description
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Not sure it makes sense to add `Malicious` under `InfoFile`

File: Integrations/integration-ATD.yml [hunk 25]
```diff
@@ -1178,10 +1201,7 @@ script:
     - contextPath: IP.Address
       description: IP's relevant to the sample
       type: string
-    - contextPath: ATD.Task.IsComplete
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Why did you remove this?

File: Integrations/integration-ATD.yml [hunk 26]
```diff
@@ -1215,8 +1235,8 @@ script:
     - contextPath: ATD.Task.submitTime
       description: Submission time of the sample
       type: string
-    - contextPath: ATD.Task.IsComplete
-      description: Stating if the detonation was complete or not
-      type: string
     description: Checks the analysis status of up to 100 jobIDs/taskIDs
   runonce: false
+releaseNotes: "Handle outputs for malicious files"
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Please elaborate more on RN

File: Integrations/integration-ATD.yml [hunk 27]
```diff
@@ -3,12 +3,13 @@ commonfields:
   version: -1
 name: McAfee Advanced Threat Defense
 display: McAfee Advanced Threat Defense
+fromversion: 3.5.0
```
Comment (Itay4) [inline Integrations/integration-ATD.yml]: Why is this?

File: Integrations/integration-ATD.yml [hunk 28]
```diff
@@ -3,12 +3,13 @@ commonfields:
...
 description: 'Integrated advanced threat detection: Enhancing protection from network
   edge to endpoint'
-detaileddescription: |
-  When entering a is a user to use the API with, please make sure the user has capability of "Allow Multiple Logins".
+detaileddescription: '  When entering a user to use the API, please make sure the
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: Done

File: Integrations/integration-ATD.yml [hunk 29]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+    ''' HELPERS '''
+    def string_arr_to_int_arr(stringArr):
+        ''' Converts a string array to int array '''
+        intArr = []
+        if isinstance(stringArr, list):
+            for index in range(len(stringArr)):
+                if stringArr[index]:
+                    intArr.append(int(stringArr[index]))
+        else:
+            intArr = [int(stringArr)]
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: removed this function from code

File: Integrations/integration-ATD.yml [hunk 30]
```diff
@@ -37,672 +38,670 @@ configuration:
...
+        return intArr
+
+    def cast_to_int(val):
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: no good reason, changing

File: Integrations/integration-ATD.yml [hunk 31]
```diff
@@ -37,672 +38,670 @@ configuration:
...
-                UNSECURE,
-                PROXY
-            );
-        if (res.StatusCode < 200 || res.StatusCode >= 300) {
-            throw 'Failed to use uri ' + BASE+uri + ', request status code: ' + res.StatusCode + ' and Body: ' + res.Body + '.';
-        }
-        var uriStr = String(uri);
-        if (uriStr.indexOf("session") === -1) {
-            logout();
+    def http_request(uri, method, headers = {}, body = {}, params={}, files={}):
```
[the remark this answers is not in the document]
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 1 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: i don't think so, each one is relevant specifically to the function beneath it

File: Integrations/integration-ATD.yml (hunk 5 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 7 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 18 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: best to write just with spaces i guess. will do

File: Integrations/integration-ATD.yml (hunk 20 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: no it does not. it will be handled in another PR

File: Integrations/integration-ATD.yml (hunk 24 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: it does not, wil lfix.

File: Integrations/integration-ATD.yml (hunk 27 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: old version is toversion 3.1.0
this version has no fromversion.

File: Integrations/integration-ATD.yml (hunk 22 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: talked with gavriel, changing type of output to string seems to be ok

File: Integrations/integration-ATD.yml (hunk 3 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: get_session is the 'atd-login' command, so no

File: Integrations/integration-ATD.yml (hunk 1 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 4 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 6 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done

File: Integrations/integration-ATD.yml (hunk 2 above)
Comment (bakatzir) [reply Integrations/integration-ATD.yml]: done
